# Oblig D: Integrasjon
## D.1

In [1]:
def rektangelmetoden(f, a, b, n):
    """
    Venstretilnærming for bestemt integrasjon med rektangler.
    f: funksjonen som skal integreres
    a: nedre intervallgrense
    b: øvre intervallgrense
    n: antall rektangler
    """
    A = 0
    h = (b-a)/n
    for k in range (n):
        A = A + f(a + k*h)
    return A*h

def trapesmetoden(f, a, b, n):
    """
    Trapesmetoden for bestemt integrasjon.
    f: funksjonen som skal integreres
    a: nedre intervallgrense
    b: øvre intervallgrense
    n: antall trapeser
    """
    h = (b-a)/n
    total = (f(a) + f(b))/2.0
    for k in range(1, n):
        total += f(a + k*h)
    return h*total

## D.2

In [10]:
import numpy as np

def f(x):
    return x*np.exp(x**2)

def f_analytisk(x):
    return np.exp(x**2)/2

N = [10, 100, 1000, 10000]
a = 1
b = 2
analytisk = f_analytisk(b) - f_analytisk(a)

for n in N:
    rekt = rektangelmetoden(f,a,b,n)
    trap = trapesmetoden(f,a,b,n)
    avvik_rekt = abs(analytisk-rekt)/analytisk * 100
    avvik_trap = abs(analytisk-trap)/analytisk * 100
    print(f'Rektangelmetoden med {n} rektangler gir: {rekt:.4f}, avvik: {avvik_rekt:.2f} %')
    print(f'Trapesmetoden med {n} trapeser gir: {trap:.4f}, avvik: {avvik_trap:.2f} %')

Rektangelmetoden med 10 rektangler gir: 21.0170, avvik: 18.98 %
Trapesmetoden med 10 rektangler gir: 26.3409, avvik: 1.55 %
Rektangelmetoden med 100 rektangler gir: 25.4116, avvik: 2.04 %
Trapesmetoden med 100 rektangler gir: 25.9440, avvik: 0.02 %
Rektangelmetoden med 1000 rektangler gir: 25.8867, avvik: 0.21 %
Trapesmetoden med 1000 rektangler gir: 25.9400, avvik: 0.00 %
Rektangelmetoden med 10000 rektangler gir: 25.9346, avvik: 0.02 %
Trapesmetoden med 10000 rektangler gir: 25.9399, avvik: 0.00 %


## D.3

In [26]:
def MC_int(f, a, b, n):
    """
    Monte Carlo-metode for bestemt integrasjon.
    f: funksjonen som skal integreres
    a: nedre intervallgrense
    b: øvre intervallgrense
    n: antall utvalgsverdier
    """
    x = np.random.uniform(a, b, n)  
    f_bar = np.sum(f(x))  
    f_int = (b-a)/n * f_bar          
    return f_int

for n in N:
    montecarlo = MC_int(f, a, b ,n)
    avvik_MC = abs(analytisk-montecarlo)/analytisk * 100
    print(f'MC-metoden med N = {n} gir: {montecarlo}, avvik: {avvik_MC:.2f} %')

MC-metoden med N = 10 gir: 18.744477247013133, avvik: 27.74 %
MC-metoden med N = 100 gir: 25.522784891447433, avvik: 1.61 %
MC-metoden med N = 1000 gir: 25.84338873846833, avvik: 0.37 %
MC-metoden med N = 10000 gir: 25.98222374381607, avvik: 0.16 %


## D.4

In [25]:
from scipy import integrate
from scipy.misc import derivative
import numpy as np

# Integrerasjon
for n in N:
    x = np.linspace(a,b,n)
    y = f(x)
    
    simpsons = integrate.simps(y,x)
    gauss_kvadratur = integrate.quad(f,a,b)
    avvik_simps = abs(analytisk-simpsons)/analytisk * 100
    avvik_kvad = abs(analytisk-gauss_kvadratur[0])/analytisk * 100
    
    print(f'Simpsons metode med {n} rektangler gir: {simpsons}, avvik: {avvik_simps:.2f} %')
    print(f'Gauss kvadratur med {n} rektangler gir: {gauss_kvadratur[0]}, avvik: {avvik_kvad:.2f} %')

Simpsons metode med 10 rektangler gir: 26.053456535604678, avvik: 0.44 %
Gauss kvadratur med 10 rektangler gir: 25.9399341023426, avvik: 0.00 %
Simpsons metode med 100 rektangler gir: 25.940036474655006, avvik: 0.00 %
Gauss kvadratur med 100 rektangler gir: 25.9399341023426, avvik: 0.00 %
Simpsons metode med 1000 rektangler gir: 25.939934203685322, avvik: 0.00 %
Gauss kvadratur med 1000 rektangler gir: 25.9399341023426, avvik: 0.00 %
Simpsons metode med 10000 rektangler gir: 25.93993410244384, avvik: 0.00 %
Gauss kvadratur med 10000 rektangler gir: 25.9399341023426, avvik: 0.00 %


## D.5

Tabell over avvik i %:

|  N               | 10 | 100 | 1000 | 10 000 |
|------------------|----|-----|------|--------|
| Rektangelmetoden | 18.98   | 2.04    | 0.21     | 0.02       |
| Trapesmetoden    |  1.55  | 0.02    |  0.00    |    0.00    |
| MC-metode        | 27.14   | 1.61     |  0.37     |  0.16       |
| Simpsons metode  |0.44    | 0.00     |  0.00    | 0.00        |
| Gauss kvadratur  |  0.00  |  0.00   |    0.00  |  0.00      |

Vi ser at trapesmetoden får raskest minst feil, men at rektangelmetoden også ikke er så verst ettersom *N* øker. MC-metoden gir aller dårligst resultat, og trenger svært høy verdi av *N* for å gi et greit resultat. Dette er fordi den er en statistisk metode, og trenger derfor mange verdier for å få en god tilnærming. MC-metoder egner seg best for multiple integraler. Den desidert mest presise metoden er Gauss kvadratur, tett etterfulgt av Simpsons metode.